jni: consider /data/app to the fd whitelist if Xposed is detected
Latest security update has added whitelisting routine to the Zygote.

Since Xposed now reads from /data/app,
it's necessary to add /data/app to the whitelist.

Dynamically do this if XposedBridge.jar is detected.

Signed-off-by: Park Ju Hyung <qkrwngud825@gmail.com>

Original commit: https://github.com/ResurrectionRemix/android_frameworks_base/commit/a12db47b6434486a81057340568dbf8e30d50114#diff-7b8d35fc4a3bb7df07f04382a5398a58

project frameworks/base/
diff --git a/frameworks/base/core/jni/fd_utils-inl.h b/frameworks/base/core/jni/fd_utils-inl.h
index f245a7f..00c0d98 100644
--- a/frameworks/base/core/jni/fd_utils-inl.h
+++ b/frameworks/base/core/jni/fd_utils-inl.h
@@ -264,6 +264,17 @@ class FileDescriptorInfo {
         path.compare(path.size() - kJarSuffix.size(), kJarSuffix.size(), kJarSuffix) == 0) {
       return true;
     }
+
+    if (access("/system/framework/XposedBridge.jar", F_OK ) != -1) {
+      // Xposed-powered Zygote might read from extensions other than .apk
+      // so skip extension check
+      ALOGW("Xposed detected, loosening up Zygote fd check!");
+      static const std::string kDataAppPrefix = "/data/app/";
+      if (path.compare(0, kDataAppPrefix.size(), kDataAppPrefix) == 0) {
+        return true;
+      }
+    }
+
     return false;
   }
 
